cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,494 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/skyrings/skyring-common/tools/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/private/testcontext"
+	"storj.io/storj/private/testrand"
+	"storj.io/storj/satellite/metainfo"
+)
+
+func TestMain(m *testing.M) {
+	rand.Seed(time.Now().UnixNano())
+	os.Exit(m.Run())
+}
+
+func TestObserver_processSegment(t *testing.T) {
+	t.Run("valid objects of different projects", func(t *testing.T) {
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		obsvr := observer{
+			objects: make(bucketsObjects),
+		}
+
+		testdata1 := generateTestdataObjects(ctx.Context, t, false, false)
+		// Call processSegment with testadata objects of the first project
+		for _, objSeg := range testdata1.objSegments {
+			err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+			require.NoError(t, err)
+		}
+
+		testdata2 := generateTestdataObjects(ctx.Context, t, false, false)
+		// Call processSegment with testadata objects of the second project
+		for _, objSeg := range testdata2.objSegments {
+			err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+			require.NoError(t, err)
+		}
+
+		// Inspect observer internal state to assert that it only has the state
+		// related to the second project
+		assert.Equal(t, testdata2.projectID.String(), obsvr.lastProjectID, "lastProjectID")
+		if assert.Equal(t, len(testdata2.expectedObjects), len(obsvr.objects), "objects number") {
+			for bucket, bucketObjs := range obsvr.objects {
+				expBucketObjs, ok := testdata2.expectedObjects[bucket]
+				if !ok {
+					t.Errorf("bucket '%s' shouldn't exist in objects map", bucket)
```

Comment:
I am confused here.

Line 59 gets the object for the specified `bucket`. `ok` is set to `true` if `bucket` exists on the map.

Then on line 60, we check if the `bucket` is missing on the map.

Then on line 61 we fail with an error that the bucket should NOT exist... Shouldn't the error say that the bucket was expected to exist on the map?

Reply:
Also, can we use `assert.Truef(ok, ...)`?

Reply:
> Then on line 61 we fail with an error that the bucket should NOT exist... Shouldn't the error say that the bucket was expected to exist on the map?

No, because it iterates on the map assigned to the observer instance; if the bucket exists on it but it doesn't exist in the expected buckets map, then that bucket shouldn't exist in the returned value.

> also, can we use assert.Truef(ok, ...)?

Yes, we could use `assert.Truef...` but we need anyway to have a conditional because we want to continue the loop in case that the assertion fails.

I'll change it.